msresist/expectation_maximization.py

```diff
@@ -25,21 +26,41 @@ def EM_clustering(data, info, ncl, seqDist=None, gmmIn=None):
 
     # Indices for looking up probabilities later.
     idxx = np.atleast_2d(np.arange(d.shape[0]))
+
+    # In case we have missing data, use SVD-EM to fill it for initialization
+    kmD = np.copy(d)
+    pc = PCA(kmD, ncomp=ncl, missing="fill-em")
+
+    # Solve for the KMeans clustering for initialization
+    km = KMeans(ncl, tol=1e-9)
+    km.fit(pc._adjusted_data)
+
+    # Add a dummy variable for the sequence information
     d = np.hstack((d, idxx.T))
 
     for _ in range(10):
         if gmmIn is None:
             # Initialize model
             dists = list()
-            for _ in range(ncl):
-                nDist = [NormalDistribution(sp.norm.rvs(), 0.2) for _ in range(d.shape[1] - 1)]
-                dists.append(IndependentComponentsDistribution(nDist + [seqDist.copy()]))
+            for ii in range(ncl):
+                nDist = [NormalDistribution(1.0, 0.2) for _ in range(d.shape[1] - 1)] + [seqDist.copy()]
+                
+                for jj in range(d.shape[1] - 1):
+                    nDist[jj].fit(d[km.labels_ == ii, jj])
+
+                weights = np.array(km.labels_ == ii, dtype=float)
+                weights = 0.9 * weights + 0.01
+
+                nDist[-1].summarize(d[:, -1], weights=weights)
+                nDist[-1].from_summaries()
+
+                dists.append(IndependentComponentsDistribution(nDist))
 
             gmm = GeneralMixtureModel(dists)
         else:
             gmm = gmmIn
 
-        gmm.fit(d, max_iterations=50, verbose=False, stop_threshold=1e-2)
+        gmm.fit(d, max_iterations=200, verbose=False, stop_threshold=1e-3)
```

Comment:
The fits seem to improve for much longer now. I think this might be from renormalizing the logP's, which reduces roundoff errors.